### PR TITLE
feat: implement opensearch serverless support

### PR DESCRIPTION
**Pull request by hvital**
Fixes # 136

This PR adds an existingOpensearchServerlessCollection option to support importing existing OpenSearch Serverless collections.

---
By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of the project license.

### CODE REVIEW

File: src/patterns/gen-ai/aws-qa-appsync-opensearch/index.ts [hunk 1]
```diff
@@ -449,27 +457,39 @@ export class QaAppsyncOpensearch extends Construct {
...
-            ':domain/' +
-            props.existingOpensearchDomain.domainName +
-            '/*',
-          'arn:aws:es:' +
-            Aws.REGION +
-            ':' +
-            Aws.ACCOUNT_ID +
-            ':domain/' +
-            props.existingOpensearchDomain.domainName,
+          'arn:aws:aoss:'+Aws.REGION+':'+Aws.ACCOUNT_ID+':collection/'+props.openSearchIndexName,
```
Comment (scottschreckengaust) [inline src/patterns/gen-ai/aws-qa-appsync-opensearch/index.ts:489]:
```suggestion
          'arn:'+Aws.PARTITION+':aoss:'+Aws.REGION+':'+Aws.ACCOUNT_ID+':collection/'+props.openSearchIndexName,
```

File: src/patterns/gen-ai/aws-qa-appsync-opensearch/index.ts [hunk 2]
```diff
@@ -449,27 +457,39 @@ export class QaAppsyncOpensearch extends Construct {
...
 
-    question_answering_function_role.addToPolicy(
-      new iam.PolicyStatement({
+    if (props.existingOpensearchDomain) {
+      question_answering_function_role.addToPolicy(
+        new iam.PolicyStatement({
+          effect: iam.Effect.ALLOW,
+          actions: ['es:*'],
+          resources: [
+            'arn:aws:es:' +
```
Comment (scottschreckengaust) [inline src/patterns/gen-ai/aws-qa-appsync-opensearch/index.ts:466]:
```suggestion
            'arn:'+Aws.PARTITION+':es:' +
```

File: src/patterns/gen-ai/aws-qa-appsync-opensearch/index.ts [hunk 3]
```diff
@@ -449,27 +457,39 @@ export class QaAppsyncOpensearch extends Construct {
...
+              Aws.REGION +
+              ':' +
+              Aws.ACCOUNT_ID +
+              ':domain/' +
+              props.existingOpensearchDomain.domainName +
+              '/*',
+            'arn:aws:es:' +
```
Comment (scottschreckengaust) [inline src/patterns/gen-ai/aws-qa-appsync-opensearch/index.ts:473]:
```suggestion
            'arn:'+Aws.PARTITION+':es:' +
```

File: src/patterns/gen-ai/aws-rag-appsync-stepfn-opensearch/index.ts [hunk 4]
```diff
@@ -636,14 +645,26 @@ export class RagAppsyncStepfnOpensearch extends Construct {
...
-        'arn:aws:es:'+Aws.REGION+':'+Aws.ACCOUNT_ID+':domain/'+props.existingOpensearchDomain.domainName+'/*',
-        'arn:aws:es:'+Aws.REGION+':'+Aws.ACCOUNT_ID+':domain/'+props.existingOpensearchDomain.domainName,
-      ],
-    }));
+    if (props.existingOpensearchDomain) {
+      embeddings_job_function_role.addToPolicy(new iam.PolicyStatement({
+        effect: iam.Effect.ALLOW,
+        actions: ['es:*'],
+        resources: [
+          'arn:aws:es:'+Aws.REGION+':'+Aws.ACCOUNT_ID+':domain/'+props.existingOpensearchDomain.domainName+'/*',
```
Comment (scottschreckengaust) [inline src/patterns/gen-ai/aws-rag-appsync-stepfn-opensearch/index.ts:653]:
```suggestion
          'arn:'+Aws.PARTITION+':es:'+Aws.REGION+':'+Aws.ACCOUNT_ID+':domain/'+props.existingOpensearchDomain.domainName+'/*',
```

File: src/patterns/gen-ai/aws-rag-appsync-stepfn-opensearch/index.ts [hunk 5]
```diff
@@ -636,14 +645,26 @@ export class RagAppsyncStepfnOpensearch extends Construct {
...
+          'arn:aws:es:'+Aws.REGION+':'+Aws.ACCOUNT_ID+':domain/'+props.existingOpensearchDomain.domainName,
```
Comment (scottschreckengaust) [inline src/patterns/gen-ai/aws-rag-appsync-stepfn-opensearch/index.ts:654]:
```suggestion
          'arn:'+Aws.PARTITION+':es:'+Aws.REGION+':'+Aws.ACCOUNT_ID+':domain/'+props.existingOpensearchDomain.domainName,
```

File: src/patterns/gen-ai/aws-rag-appsync-stepfn-opensearch/index.ts [hunk 6]
```diff
@@ -636,14 +645,26 @@ export class RagAppsyncStepfnOpensearch extends Construct {
...
+        ],
+      }));
+    }
+
+    if (props.existingOpensearchServerlessCollection) {
+      embeddings_job_function_role.addToPolicy(new iam.PolicyStatement({
+        effect: iam.Effect.ALLOW,
+        actions: ['aoss:APIAccessAll'],
+        resources: [
+          'arn:aws:aoss:'+Aws.REGION+':'+Aws.ACCOUNT_ID+':collection/'+props.openSearchIndexName,
```
Comment (scottschreckengaust) [inline src/patterns/gen-ai/aws-rag-appsync-stepfn-opensearch/index.ts:664]:
```suggestion
          'arn:'+Aws.PARTITION+':aoss:'+Aws.REGION+':'+Aws.ACCOUNT_ID+':collection/'+props.openSearchIndexName,
```

File: test/patterns/gen-ai/aws-rag-appsync-stepfn-opensearch/aws-rag-appsync-stepfn-opensearch-serverless.test.ts [hunk 7]
```diff
@@ -0,0 +1,198 @@
...
+  let ragTestConstruct: RagAppsyncStepfnOpensearch;
+  const cognitoPoolId = 'region_XXXXX';
+
+  afterAll(() => {
+    console.log('Test completed');
+  });
+
+  beforeAll(() => {
+    const ragTestStack = new cdk.Stack(undefined, undefined, {
+      env: { account: cdk.Aws.ACCOUNT_ID, region: cdk.Aws.REGION },
```
Comment (scottschreckengaust) [inline test/patterns/gen-ai/aws-rag-appsync-stepfn-opensearch/aws-rag-appsync-stepfn-opensearch-serverless.test.ts:34]: Is it possible for this to be a snapshot instead from a mocked or fictitious account and region?

https://github.com/awslabs/generative-ai-cdk-constructs/blob/2a8188a3122c7c6c581d54fd74aab01e53f688f2/test/common/helpers/utils.test.ts#L31C5-L31C119

File: src/patterns/gen-ai/aws-qa-appsync-opensearch/index.ts (hunk 1 above)
Comment (hvital) [reply src/patterns/gen-ai/aws-qa-appsync-opensearch/index.ts]: [refactor: added partition dynamic value](https://github.com/awslabs/generative-ai-cdk-constructs/pull/212/commits/99f2ed5e738a645fd008aa2172982e4a173d7893)

File: src/patterns/gen-ai/aws-qa-appsync-opensearch/index.ts (hunk 2 above)
Comment (hvital) [reply src/patterns/gen-ai/aws-qa-appsync-opensearch/index.ts]: [refactor: added partition dynamic value](https://github.com/awslabs/generative-ai-cdk-constructs/pull/212/commits/99f2ed5e738a645fd008aa2172982e4a173d7893)

File: src/patterns/gen-ai/aws-qa-appsync-opensearch/index.ts (hunk 3 above)
Comment (hvital) [reply src/patterns/gen-ai/aws-qa-appsync-opensearch/index.ts]: [refactor: added partition dynamic value](https://github.com/awslabs/generative-ai-cdk-constructs/pull/212/commits/99f2ed5e738a645fd008aa2172982e4a173d7893)

File: src/patterns/gen-ai/aws-rag-appsync-stepfn-opensearch/index.ts (hunk 4 above)
Comment (hvital) [reply src/patterns/gen-ai/aws-rag-appsync-stepfn-opensearch/index.ts]: [refactor: added partition dynamic value](https://github.com/awslabs/generative-ai-cdk-constructs/pull/212/commits/99f2ed5e738a645fd008aa2172982e4a173d7893)

File: src/patterns/gen-ai/aws-rag-appsync-stepfn-opensearch/index.ts (hunk 5 above)
Comment (hvital) [reply src/patterns/gen-ai/aws-rag-appsync-stepfn-opensearch/index.ts]: [refactor: added partition dynamic value](https://github.com/awslabs/generative-ai-cdk-constructs/pull/212/commits/99f2ed5e738a645fd008aa2172982e4a173d7893)

File: src/patterns/gen-ai/aws-rag-appsync-stepfn-opensearch/index.ts (hunk 6 above)
Comment (hvital) [reply src/patterns/gen-ai/aws-rag-appsync-stepfn-opensearch/index.ts]: [refactor: added partition dynamic value](https://github.com/awslabs/generative-ai-cdk-constructs/pull/212/commits/99f2ed5e738a645fd008aa2172982e4a173d7893)

**Review by scottschreckengaust — APPROVED** [review]
LGTM, next PR increase the code coverage

**Review by dineshSajwan — APPROVED** [review]
LGTM

**Pull request merged**